pkg/apis/allocation/v1/gameserverallocation.go

```diff
@@ -174,9 +212,84 @@ func (s *GameServerSelector) Matches(gs *agonesv1.GameServer) bool {
 		}
 	}
 
+	if runtime.FeatureEnabled(runtime.FeatureCountsAndLists) {
+		// Only check for matches if there are CounterSelectors or ListSelectors
+		if (s.Counters != nil) && (len(s.Counters) != 0) {
+			if !(s.matchCounters(gs)) {
+				return false
+			}
+		}
+		if (s.Lists != nil) && (len(s.Lists) != 0) {
+			if !(s.matchLists(gs)) {
+				return false
+			}
+		}
+	}
+
 	return true
 }
 
+// matchCounters returns true if there is a match for the CounterSelector in the GameServerStatus
+func (s *GameServerSelector) matchCounters(gs *agonesv1.GameServer) bool {
+	if s.Counters != nil && gs.Status.Counters != nil {
+		for counter, counterSelector := range s.Counters {
+			// If the Counter Selector does not exist in GameServerStatus, return false.
+			counterStatus, ok := gs.Status.Counters[counter]
+			if !ok {
+				return false
+			}
+			// 0 means undefined (unlimited) for MaxAvailable.
+			available := counterStatus.Capacity - counterStatus.Count
+			if available < counterSelector.MinAvailable ||
+				(counterSelector.MaxAvailable != 0 && available > counterSelector.MaxAvailable) {
+				return false
+			}
+			// 0 means undefined (unlimited) for MaxCount.
+			if counterStatus.Count < counterSelector.MinCount ||
+				(counterSelector.MaxCount != 0 && counterStatus.Count > counterSelector.MaxCount) {
+				return false
+			}
+		}
+		return true
+	}
+	return false
```

Comment:
nitpick: To improve readability, this blob can be rewritten like:
~~~
if gs.Status.Counters == nil {
    return false
}

for counter, counterSelector : range s.Counters {
....
}

return true
~~~

No need to do `if s.Counters != nil` again since it's already checked.